File: leetcode/medium/daily_temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        N = len(temperatures)
        ans = [0] * N
        stack = []
        for i,t in enumerate(temperatures):
            while stack and t > temperatures[stack[-1]]:
                sc = stack[-1]
                ans[sc] = i - sc
                stack.pop()
            stack.append(i)
        return ans

File: leetcode/medium/dailyTemperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& T) {
        int N = T.size();
        vector<int> stack, ans(N);
        for (int i = 0; i < N; ++i) {
            int t = T[i];
            while (!stack.empty() && t > T[stack.back()]) {
                int sb = stack.back();
                ans[sb] = i - sb;
                stack.pop_back();
            }
            stack.push_back(i);
        }
        return ans;
    }
};
